proposals/unsigned/unsigned.txt

```diff
@@ -0,0 +1,228 @@
+inclusion in Fortran 202y.
+
+The use cases can be roughly divided into three classes:
+
+- representing unsigned integer
+- bit operations
+- modular arithmetic (2^n for a datatype with n bits)
+
+The two fundamental designs are:
+
+- adding a dedicated type for each use case with the appropriate
+  behavior of aritmetic operators and intrinsic functions on overflow;
+  the different types can possibly just be different kinds for
+  `unsigned(kind=...)`:
+  * `unsigned`: arithmetic operation overflow do not wraparound, or are
+    possibly not even defined
+  * `bits`: bit operations are defined, arithmetic operations do not
+    wraparound or are not defined
+  * `modular`: arithmetic operations wraparound using modular 2^n
+    arithmetic
+- one `unsigned` type that is used for all three use cases; intrinsic
+  functions are used to implement bit operations, various overflow modes
+  (wraparound, checked, saturated), etc. One must choose some default
+  behavior on arithmetic overflow, discussed below.
+
+There is currently no community nor committee agreement which of the two
+fundamnetal designs to do, nor what the default overflow behavior should
+be for arithmetic operations if we go with the second design.
+
+Consequently, we are proposing to implement the second design with
+undefined behavior for arithmetic overflow, consistent with the existing
+signed integers in Fortran, which allows processors to optionally check
+for overflow. This proposal leaves the door open to later implement
+either the first design, or the second design with defined overflow
+behavior (to wraparound). It is also a subset of features that most
+people seem to agree that we need.
+
+The proposal adds a solution to all three use cases (data
+representation, bit operations, modular arithmetic) that processors can
+implement and users can start using. If later we decide to either add
+dedicated types/kinds `bits` and `modular`, or define default arithmetic
+operators' overflow to wraparound, no existing code will break.
+
+## 1.1. Prior art
+
+At least one Fortran compiler, Sun Fortran, supported unsigned integers.
+Documentation can be found at [Oracle]
+(https://docs.oracle.com/cd/E19205-01/819-5263/aevnb/index.html).
+This proposal borrows heavily from that prior art, without sticking
+to it in all details.
+
+## 1.2 Inputs to this proposal
+
+In addition to the references listed above, the discussion at the
+Fortran proposals site
+https://github.com/j3-fortran/fortran_proposals/issues/2
+influenced this proposal.
+
+
+# 2. Goal
+
+Define a new type, UNSIGNED, with a small set of intrinsic operations
+and intrinsic functions that would satisfy most of the use cases listed
+above.
+
+## 2.1 Value range limitation
+
+An UNSIGNED with n bits has a value range between 0 and 2^n-1.
+(Note that Fortran model integers have values between -2^(n-1)+1 and
+2^(n-1)-1).
+
+## 2.2 Arithmetic overflow is undefined
```

Comment:
This section is contrary to actual hardware behavior (truncation), 50+ years of expectations from C/C++, and current implementations of UNSIGNED (Sun, gfortran 15, LLVM Flang, and XLF's unsigned vector elements).

Reply:
This proposal still leaves the door open to later add default wraparound on overflow. 

(Note that you literally just implemented a prototype yesterday into LLVM Flang as a Draft PR: https://github.com/llvm/llvm-project/pull/113504, so I don't know how much this qualifies as "current implementation".)

Reply:
By the time your glacial process gets around to making any kind of progress, it will have been available for years.

Weren't you the guy who wanted the committee to validate concepts with prototypes?

Reply:
> By the time your glacial process gets around to making any kind of progress, it will have been available for years.

It's not my process and I don't agree with it. I tried to change things, including this very repository or running for committee's leadership or starting a compiler. I don't know what else you want me to do.

> Weren't you the guy who wanted the committee to validate concepts with prototypes?

Yes, I've been pushing for that for years. Every proposal needs to have a prototype.

But you should implement the right prototype, and try to reach a consensus with it. So far you implemented a design that *you* like, and are hoping that if LLVM Flang is successful it will become the de-facto standard. That is no different to how the committee has been operating, not always, but sometimes the decision is effectively done by one person without input from others on the committee or wider community. You rightfully object to that, and I agree. But if your solution to this is problem is to simply push things to LLVM Flang because you have been one of the main authors, without feedback from others, then this is no better than what the committee does, and I object to that too.

Reply:
> But you should implement the right prototype, and try to reach a consensus with it. So far you implemented a design that _you_ like, and are hoping that if LLVM Flang is successful it will become the de-facto standard.

Note that LLVM is not alone in implementing this, this is coordinated (and compatible) with what gfortran does.

Reply:
> So far you implemented a design that you like, and are hoping that if LLVM Flang is successful it will become the de-facto standard.

It bears repeating that neither GNU Fortran nor LLVM Flang's unsigned features will be available by default.

There is precedent in C and descendants, prior Fortran extensions, existing hardware, and, in my judgement, an informed consensus on this subject.  The arguments against guaranteeing the truncating behavior were not convincing to me (future hardware might do something other than truncate?  optimizations might be inhibited? user code can't check for overflow?), and leaving behavior undefined or compiler-dependent undermines the goal of portability that is the whole reason for having standards in the first place.

So write your proposal is a completely defined manner.  Don't put undefined or compiler-dependent behavior into the definition of a new feature -- that's only (marginally) justifiable when standardizing parts of an existing feature that have incompatible and unchangeable implementations already present.  If you don't want truncating semantics, then put whatever you *do* want into your proposal.

Reply:
@klausler those are good points.

> The arguments against guaranteeing the truncating behavior were not convincing to me (future hardware might do something other than truncate? optimizations might be inhibited? user code can't check for overflow?)

These are not convincing arguments for me either. The arguments against modular arithmetic by default is that:

* assuming modular arithmetic is a minor (but important) use case (I know we don't have an agreement on this assumption)
* and consequently that the majority of use cases are data representation and bit operations (all the use cases that do not require modular arithmetic, including interoperability with C, etc.)
* given that often you might get a size argument as unsigned integer (say from some C system API or a C++ code where it is a habit to use unsigned for sizes)
* you do not arbitrarily add modulo operation in arithmetic, so that you can safely use `size - 1` and other common operations. If you make a mistake, the compiler will tell you, instead of being required to wraparound and not tell you.

If you do want to use modular arithmetic, you use `add_wrapping` (name to be decided) or later you can have a dedicated type just for that, if desired.

Reply:
> leaving behavior undefined or compiler-dependent undermines the goal of portability that is the whole reason for having standards in the first place.

This is not true. The assignment `a(1:n) = b(1:m)` has an undefined behavior for `n /= m`: this is illegal code, but the compiler is not required to detect and report the violation. This is intentional to maximize the performances of the legal codes, and it doesn't undermine the portability. If this assignment was legal or if the compiler was required to report the violation, the compiler would have to make additional runtime checks (comparing n to m in this example).

Reply:
> > The arguments against guaranteeing the truncating behavior were not convincing to me (future hardware might do something other than truncate? optimizations might be inhibited? user code can't check for overflow?)
> 
> These are not convincing arguments for me either (except the last one).

These are nonetheless valid arguments for a language that often tries abstracting the hardware.

Reply:
> > leaving behavior undefined or compiler-dependent undermines the goal of portability that is the whole reason for having standards in the first place.
> 
> This is not true. The assignment `a(1:n) = b(1:m)` has an undefined behavior for `n /= m`: this is illegal code, but the compiler is not required to detect and report the violation. This is intentional to maximize the performances of the legal codes, and it doesn't undermine the portability. If this assignment was legal or if the compiler was required to report the violation, the compiler would have to make additional runtime checks (comparing n to m in this example).

There are distinctions to be drawn between undefined behavior, non-conforming errors that may be undetected by the compiler, and implementation-dependent behavior.  Some of these do not contribute to portability across compilers when they appear in a language standard.  It is better for a standard to mandate a specific behavior that can be observed and validated with readily available tests.  This is especially true for features that map directly to universally available hardware semantics.

Reply:
> * you do not arbitrarily add modulo operation (wraparound) in arithmetic, so that you can safely use `size - 1` and other common operations. If you make a mistake, the compiler will tell you, instead of being required to wraparound and not tell you.

I think that you're assuming two things here: that compilers will always catch over/underflow in constant expression if the language doesn't define the behavior to be modular, and that they can't support options to emit a warning in those cases if the language does define the behavior to be modular.

Reply:
> that compilers will always catch over/underflow in constant expression if the language doesn't define the behavior to be modular

I am assuming the compiler *can* always catch it, and indeed that seems to be possible to implement, if you play with such expressions in Rust, you will always get either compile-time or runtime errors. However in "Release" mode users would typically disable such checks (unlike in Rust where I think it's not possibly to disable it). Also some compilers can decide not to implement such checks and it would not break anything.

> that they can't support options to emit a warning in those cases if the language does define the behavior to be modular.

The underflow `size-1` check is a fundamentally runtime check. If we standardize that `size-1` must wraparound, how can the compiler insert runtime checks for this? The KISS example would stop working then. I think it is not possible for the compiler to tell if you want modular arithmetic or not if the wraparound is standardized. Please correct me if I am wrong.

Reply:
> The underflow `size-1` check is a fundamentally runtime check. If we standardize that `size-1` must wraparound, how can the compiler insert runtime checks for this?

Introduce a derived type like the `cu`below.
```
module checked_unsigned
  implicit none
  private
  public :: uc, operator(+), assignment(=)
  type uc
     unsigned :: u
  end type uc
  interface operator(+)
     procedure checked_add_uc
     procedure checked_add_cu
     procedure checked_add_cc
  end interface operator(+)
  interface assignment(=)
     procedure set_uc
     procedure set_cu
  end interface assignment(=)
contains
  recursive elemental function checked_add_uc(a,b) result(c)
    unsigned, intent(in) :: a
    type(uc), intent(in) :: b
    type(uc) :: c
    c%u = a + b%u
    if (c%u < a) error stop
  end function checked_add_uc
  recursive elemental function checked_add_cu(a,b) result(c)
    type(uc), intent(in) :: a
    unsigned, intent(in) :: b
    type(uc) :: c
    c%u = a%u + b
    if (c%u < a%u) error stop
  end function checked_add_cu
  recursive elemental function checked_add_cc(a,b) result(c)
    type(uc), intent(in) :: a, b
    type(uc) :: c
    c%u = a%u + b%u
    if (c%u < a%u) error stop
  end function checked_add_cc
  recursive elemental subroutine set_uc(a,b)
    type(uc), intent(in) :: b
    unsigned, intent(out) :: a
    a = b%u
  end subroutine set_uc
  recursive elemental subroutine set_cu(a,b)
    unsigned, intent(in) :: b
    type(uc), intent(out) :: a
    a%u = b
  end subroutine set_cu
end module checked_unsigned

program main
  use checked_unsigned
  implicit none
  type(uc) :: a
  a = huge(a%u)
  print *,a + 1u
end program main

```
That took me a about 10 minutes to type up, because Fortran is a verbose language. The above could also easily be extended as a PDT, which I didn't fell like doing tonight.

To be really effective for checking, this could use two intrinsics: One more easily check for overflow (that would also greatly help integers) and something like gcc_unreachable, let's say `UNREACHABLE()`. Depending on the settings, the compiler could ignore it completely, make optimizations on the assumption that it will never be reached, or treat it as an `ERROR STOP`. This could even be an intrinsic module for people who are concerned about users getting it wrong.

And my point of laying traps for people who are used to C still stands. Yes, we have deviations from what C does, but these are caught at compile-time. A compiler doing time-travel optimization based on the assumption that overflow cannot happen will bite people, and believe me - that will happen.

You keep bringing up Rust. Rust is a bad example - Fortran is not rust, its behavior of trapping on debug and prescrbing wrapping on non-debug builds is silly, and Fortran should never be allowed to use unsigned integers for array indices, like Rust does. Everybody has to convert to an `INTEGER` anyway before using an `UNSIGNED`  as an array index or for a character substring, so the danger of users falling into that particular trap is not something that I am concerned about at all.  Everything which counts things in intrinsics etc is left well alone, for that reason.

Reply:
> You keep bringing up Rust. Rust is a bad example - Fortran is not rust,

And you keep bringing up C ("Fortran unsigned should be modular becasue C unsigned is"), but Fortran is not C. Whether the Rust approach isgood or not, at least they didn't try just copying C.


> Introduce a derived type like the `cu`below [...] That took me a about 10 minutes to type up,

It took you 10mn because you just wrote the addition part. One also need the subraction, the multiplication, the division, and the PDT versions of all the intrisics that work with `unsigned`. Granted, once written it's OK (*). But if we want to avoid everyone reinventing the wheel by writing their own version (and incompatible with each other version) **it really should be an intrinsic module** `iso_checked_unsigned` that provide `type(uc)` and all the needed routines. Then at this point:
- This is now very close to natively providing two intrinsic types `modular unsigned` and `unchecked unsigned`, that was the option 1 of @certik. Why having only one of them as intrinsic and the other one as derived type ?
- Why not the opposite, the intrinsic type being the (un)checked one, and the derived type being for the wrapped behavior?

About the performances: what you wrote above is a derived type that **always** check the overflows, which is not what people want most of time. One need something as fast as possible in "release mode", without any check, and possibly enabled checks by the compiler in debug mode. For that you are proposing new intrinsics that would be dedicated to overflow checking (and that should be standardized first, which is definitely not granted), and that would be active or not depending on some compiler switchs... Now it gets extremely confusing IMO, especially if it's buried inside an intrinsic module: one has to choose between a type with checked overflows or unchecked overflows, it cannot be "processor dependent".

(*) which doesn't mean it's as convenient as an intrisic type anyway... e.g. it's not possible to directly write a literal constant with a derived type.